opentelemetry-api/src/opentelemetry/trace/span.py

```diff
@@ -146,10 +154,134 @@ class TraceState(typing.Dict[str, str]):
         https://www.w3.org/TR/trace-context/#tracestate-field
     """
 
+    def __init__(
+        self,
+        entries: typing.Optional[
+            typing.Sequence[typing.Tuple[str, str]]
+        ] = None,
+    ) -> None:
+        self._dict = collections.OrderedDict()
```

Comment:
I wonder if we should keep the previous behaviour and have `TraceState` be the actual dictionary, instead of having to construct a `TraceState` object?

Reply:
That would make the `TraceState` to be mutable by anyone and insertion order remembering is not a language feature until `python3.7`. Specs says it should be immutable list of string key/value pairs, and all mutable operations (through add, update, delete) should validate the inputs and return a new `TraceState` and order of pairs should always be preserved.

_API should provide the `add`, `update`, `delete` [operations](https://www.w3.org/TR/trace-context/#mutating-the-tracestate-field)_. To achieve this we would still need to inherit from mapping and implement them. In case if we consider to drop support for lower python versions we should override `__setitem__` and `__delitem__` to not allow mutating directly. Keeping all this in mind, I think we should have `TraceState` object and it wouldn't be feasible to follow spec with actual dictionary.